cpp/src/arrow/flight/client.cc

```diff
@@ -132,29 +132,66 @@ class GrpcClientAuthReader : public ClientAuthReader {
       stream_;
 };
 
-class FlightIpcMessageReader : public ipc::MessageReader {
+// The next two classes are intertwined. To get the application
+// metadata while avoiding reimplementing RecordBatchStreamReader, we
+// create an ipc::MessageReader that is tied to the
+// MetadataRecordBatchReader. Every time an IPC message is read, it updates
+// the application metadata field of the MetadataRecordBatchReader. The
+// MetadataRecordBatchReader wraps RecordBatchStreamReader, offering an
+// additional method to get both the record batch and application
+// metadata.
+
+class GrpcIpcMessageReader;
+class GrpcStreamReader : public FlightStreamReader {
  public:
-  FlightIpcMessageReader(std::unique_ptr<ClientRpc> rpc,
-                         std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream)
-      : rpc_(std::move(rpc)), stream_(std::move(stream)), stream_finished_(false) {}
+  GrpcStreamReader();
+
+  static Status Open(std::unique_ptr<ClientRpc> rpc,
+                     std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream,
+                     std::unique_ptr<GrpcStreamReader>* out);
+  std::shared_ptr<Schema> schema() const override;
+  Status ReadNext(std::shared_ptr<RecordBatch>* out) override;
```

Comment:
Sorry to phone in from the peanut gallery, but what do you think about introducing a result struct like

```
struct FlightStreamChunk {
  std::shared_ptr<RecordBatch> data;
  std::shared_ptr<Buffer> app_metadata;
};
```

Reply:
Sure, that's a little more convenient for callers. We can't change the signature of `ReadNext` here though, only `ReadWithMetadata`.

Reply:
In https://github.com/apache/arrow/pull/4483 I introduced an `Iterator<T>` concept -- it's not clear to me how important it is to conform to the `RecordBatchReader` interface (or we could also use virtual inheritance so that the flight data reader can by dynamically-casted to `Iterator<std::shared_ptr<RecordBatch>>`) cc @pitrou

Reply:
Hmm, maybe it's not so important, so long as in Python it still exposes the utility methods people expect (we find it useful to do `do_get().read_pandas()` as that's by far the common case). If we could expose both iterator-of-record-batch and iterator-of-chunk that would be best (as not everyone will want the metadata).

Reply:
On the C++ side having a intended API endpoint would probably the easiest thing (since the metadata can just be ignored if the developer-user doesn't care), and we can make the Python API as ergonomic as desired (I presume -- it is our intent in fact -- that most consumers of the Python API are going to be sitting behind a front-end interface to pyarrow anyway)

Reply:
Alright, I went ahead and changed it to use that interface; when the other PR is merged, we can then make it actually implement `Iterator`.